File: KthLargestElementInAnArray.java
/**
 * LeetCode 215 Kth Largest Element in an Array
 */
class Solution {
    public int findKthLargest(int[] nums, int k) {

        PriorityQueue<Integer> priorityQueue = new PriorityQueue<>(new Comparator<Integer>() {
            @Override
            public int compare(Integer a, Integer b) {
                return a - b;
            }
        });

        for (int num : nums) {
            priorityQueue.add(num);
            if (priorityQueue.size() > k) {
                priorityQueue.poll();
            }
        }


        return priorityQueue.poll();
    }
}

File: KthLargestElementInAnArray.go
func findKthLargest(nums []int, k int) int {
	l := 0
	r := len(nums) - 1
	target := len(nums) - k
	for (l < r) {
		mid := quickSelection(nums,l,r)
		if (mid == target) {
			return nums[mid]
		}
		if (mid < target) {
			l = mid + 1
		} else {
			r = mid - 1
		}
	}
	return nums[l]
}

func quickSelection(nums[] int,l int,r int) int {
	i := l + 1
	j := r
	for {
		for (i < r && nums[i] <= nums[l]) {
			i++
		}
		for (l < j && nums[j] >= nums[l]) {
			j--
		}
		if (i >= j) {
			break
		}
		nums[i],nums[j] = nums[j],nums[i]
	}
	nums[l],nums[j] = nums[j],nums[l]
	return j
}

